packages/frontend/component/src/hooks/use-resize-observer.ts

```diff
@@ -0,0 +1,74 @@
+import type { ResizeObserverEntry } from '@juggle/resize-observer';
+import { useCallback, useEffect, useRef } from 'react';
+
+type UseResizeObserverCallback = (entity: ResizeObserverEntry) => void;
+
+let _resizeObserver: ResizeObserver | null = null;
+const elementsMap = new WeakMap<Element, UseResizeObserverCallback>();
+
+// for debugging
+(window as any)._resizeObserverElementsMap = elementsMap;
+
+/**
+ * @internal get or initialize the ResizeObserver instance
+ */
+const getResizeObserver = () =>
+  (_resizeObserver ??= new ResizeObserver(entries => {
+    entries.forEach(entry => {
+      const resizeCallback = elementsMap.get(entry.target);
+      resizeCallback?.(entry);
+    });
+  }));
+
+/**
+ * @internal unobserve an element
+ */
+const unobserve = (element: Element | null) => {
+  if (!element) return;
+  const observer = getResizeObserver();
+  observer.unobserve(element);
+  elementsMap.delete(element);
+};
+
+/**
+ * A hook that observes the resize of an element.
+ * @returns A function that can be used to observe an element.
+ *
+ * ```ts
+ * const elRef = useRef<HTMLDivElement>(null);
+ * const observe = useResizeObserver();
+ *
+ * useEffect(() => {
+ *  const dispose = observe(elRef.current, (entry) => {
+ *    console.log(entry.contentRect);
+ *  });
+ *
+ *  return () => dispose();
+ * }, [observe]);
+ * ```
+ */
+export const useResizeObserver = () => {
+  const prevElRef = useRef<Element | null>(null);
+
+  const observe = useCallback(
+    (element: Element, callback: UseResizeObserverCallback) => {
+      unobserve(prevElRef.current);
+
+      const observer = getResizeObserver();
+      prevElRef.current = element;
+      observer.observe(element);
+      elementsMap.set(element, callback);
+
+      return () => unobserve(element);
+    },
+    []
+  );
+
+  useEffect(() => {
+    return () => {
+      unobserve(prevElRef.current);
+    };
```

Comment:
If the same element being observed twice, unobserve at one place will unobserve all.

Reply:
If same element being observed for the second time, the first should be unobserved first. All of this is based on each element having only one handler.